options/posix/generic/pthread-stubs.cpp

```diff
@@ -136,8 +136,56 @@ int pthread_equal(pthread_t t1, pthread_t t2) {
 	return 0;
 }
 
+namespace {
+	struct key_global_info {
+		bool in_use;
+
+		void (*dtor)(void *);
+		uint64_t generation;
+	};
+
+	struct key_local_info {
+		void *value;
+		uint64_t generation;
+	};
+
+	constinit frg::array<
+		key_global_info,
+		PTHREAD_KEYS_MAX
+	> key_globals_{};
+
+	thread_local constinit frg::array<
+		key_local_info,
+		PTHREAD_KEYS_MAX
+	> key_locals_{};
+
+	FutexLock key_mutex_;
+}
+
 int pthread_exit(void *ret_val) {
-	auto self = static_cast<Tcb *>(mlibc::get_current_tcb());
+	auto self = mlibc::get_current_tcb();
+
+	__atomic_fetch_or(&self->cancelBits, tcbExitingBit, __ATOMIC_RELAXED);
```

Comment:
Shouldn't we check before setting this if it is already set?

Reply:
Does that matter? AFAIU if it's already set nothing will happen when we set it here.